src/computability/DFA.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+-/
+
+import data.fintype.basic
+
+/-!
+# Deterministic Finite Automata
+This file contains the definition of a Deterministic Finite Automaton (DFA), a state machine which
+determines whether a string (implemented as a list over an arbitrary alphabet) is in a regular set
+in linear time.
+-/
+
+universes u v
+
+/-- A DFA is a set of states (`state`), a transition function from state to state labelled by the
+  alphabet (`step`), a starting state (`start`) and a set of acceptance states (`accept`). -/
+structure DFA (alphabet : Type u) (state : Type v) :=
+(step : state → alphabet → state)
+(start : state)
+(accept : finset state)
```

Comment:
```suggestion
(accept : set state)
```
Then you can drop all the decidability assumptions.

More seriously, the current definition of DFA is not a *finite* automaton.  Consider e.g. the following automaton which accepts a well-known non-regular language:
```lean
def dyck : DFA bool ℤ :=
{ step := λ s a, if a then s + 1 else s - 1,
  start := 0,
  accept := {0} }
```

1. One (technically slightly inconvenient) solution would be to add a `[fintype σ]` argument to `DFA` and `NFA` (i.e. `(α : Type u) (σ : Type v) [fintype σ]`).
2. The other option would be to make this clear in the docstring, i.e. that the definition allows infinite automata, and that you need `fintype σ` if you want an actual DFA/NFA.

(Though this PR doesn't have any results that require finiteness.)

Reply:
I have discussed this with @eric-wieser and we came to the conclusion that adding a `fintype` was impractical, I will add a comment into the docstrings.